### ISSUE TITLE
Unexpected values in grouping produced by rand()

### ISSUE THREAD

**Issue by grzegorz-aniol** [neo4j/neo4j#13296]
## Issue
I observed that `rand()` function may generates sometimes unexpectedly, new values, used for grouping in another query part, separated by `WITH` clause. 

### Original query

The query doesn't fetch any data from the database and it produces regular polygons at random locations.
It has two input parameters: the number of polygons to generate and the number of points per each polygon. Increasing number of points eventually we diverge into a circle.
I reduced the complexity of the original query to the minimum to explain the issue.
Also, for simplicity, each polygon has the same size. It can be inscribed into a circle with a diameter of 1.
With the following query, I noticed quite an annoying issue that produced a shifted location of some random polygons.

```cypher
with 22 as points_in_polygon, 24 as polygons
unwind range(1, polygons) as polygon_num
with polygon_num, points_in_polygon, 20*rand() as longitude, 20*rand() as latitude
unwind range(0, points_in_polygon - 1) as pidx
with polygon_num, latitude + sin(2*3.14 * pidx/points_in_polygon) as point_latitude, longitude + cos(2*3.14 * pidx/points_in_polygon) as point_longitude
return polygon_num, count(*) as num_points, collect([point_latitude, point_longitude]) as points
```

Here is the picture of similar polygons rendered on the map:
<img src="https://github.com/neo4j/neo4j/assets/22835413/c4b1a322-e285-4390-9d02-ab1a67201baa" width="75%"/>

### Diving into details

You may think that something is wrong with calculation, but the calculation is correct. If you take a look on points generated  every 6th polygon, you will see that few of them are calculated based on completely different centroid. If you check the distance between the first and last generated point, you can check that 6th, 12th, 18th (and so on) polygon has broken last few points. The distance is much higher than expected (see other points)

```cypher
with 22 as points_in_polygon, 24 as polygons
unwind range(1, polygons) as polygon_num
with polygon_num, points_in_polygon, 20*rand() as longitude, 20*rand() as latitude
unwind range(0, points_in_polygon - 1) as pidx
with polygon_num, points_in_polygon, latitude + sin(2*3.14 * pidx/points_in_polygon) as point_latitude, longitude + cos(2*3.14 * pidx/points_in_polygon) as point_longitude
with polygon_num, points_in_polygon, count(*) as num_points, collect(point({longitude: point_latitude, latitude: point_longitude})) as points
return polygon_num, num_points, points[0], last(points), point.distance(points[0], last(points))
```

Result of the query:
<img src="https://github.com/neo4j/neo4j/assets/22835413/c0901c61-b3ed-4b4b-ab2c-44245de9508b"/>

### Into the conclusion

I explicitly grouped results by longitude, latitude. Theoretically, there should be just one group per each polygon. But every 6th polygon additional longitude and latitude is generated (wth?!) 

```cypher
with 22 as points_in_polygon, 24 as polygons
unwind range(1, polygons) as polygon_num
with polygon_num, points_in_polygon, 20*rand() as longitude, 20*rand() as latitude
unwind range(0, points_in_polygon - 1) as pidx
with polygon_num, longitude, latitude, latitude + sin(2*3.14 * pidx/points_in_polygon) as point_latitude, longitude + cos(2*3.14 * pidx/points_in_polygon) as point_longitude
return polygon_num, longitude, latitude, count(*) as num_points, collect([point_latitude, point_longitude]) as points
```

Result of the query:
![Screenshot from 2023-09-11 23-01-58](https://github.com/neo4j/neo4j/assets/22835413/dfd3e00d-8608-4543-8077-cee8112507e8)

And finally, with the following query you can see that the root cause of the query is rand() function. If you replace it with pseudo generated values, there is no any issue. 

```cypher
with 22 as points_in_polygon, 24 as polygons
unwind range(1, polygons) as polygon_num
with polygon_num, points_in_polygon, 20*0.133*polygon_num as longitude, 20*0.2345*polygon_num as latitude
unwind range(0, points_in_polygon - 1) as pidx
with polygon_num, longitude, latitude, latitude + sin(2*3.14 * pidx/points_in_polygon) as point_latitude, longitude + cos(2*3.14 * pidx/points_in_polygon) as point_longitude
return polygon_num, longitude, latitude, count(*) as num_points, collect([point_latitude, point_longitude]) as points
```

Result of the query:
![Screenshot from 2023-09-11 23-03-46](https://github.com/neo4j/neo4j/assets/22835413/cfce5422-1092-45f7-a395-ac17f7144970)

You can experiment with different number of points in polygons, e.g. 5 or 6. But then, in order to see the issue, you have to generate more polygons. 

<img src="https://github.com/neo4j/neo4j/assets/22835413/6351d62e-f713-4da7-87f1-6ad94a41ce02" width="75%"/>

## Expected result

I would expect that there will be no any additional values used for grouping, especially that values produced by `rand()` function are generated in the first part of the query, and passed to the another one in variables. Therefore it's surprising that in unexpected places some new values are implicitly fetched for grouping. 

## Neo4j version
I observe the issue on Neo4j 5.10.0 but also with 4.4.14

**Comment by AzuObs** [comment]
Hi @grzegorz-aniol,

Thank you for submitting this very detailed report to us. It has been sent to triaging by one of our Cypher teams, and hopefully will be investigated soon.

Have a pleasant day,
Daniel

**Comment by sherfert** [comment]
Hi @grzegorz-aniol 

You found indeed a bug, thanks again for the detailed report.

While we work on fixing this, it seems there is a workaround:
If you prepend your query with `CYPHER runtime=slotted` you should get correct results.

Best regards,
Satia

**Comment by grzegorz-aniol** [comment]
I'm glad I could help :D 
I confirm, the workaround solved the problem, thank you :)

**Comment by pontusmelke** [comment]
@grzegorz-aniol thanks again for reporting this. A fix has been merged in 5.13 and we will also have a fix in the next release of 4.4.

I'll close this issue now, don't hesitate to repopen or open a new issue should the problem come back.

Best regards,
Pontus
